### DISCUSSION TITLE
implementing password_auth

### DISCUSSION THREAD

**Discussion by dave42w** [maxcountryman/axum-login#125] [Ideas]
Hi,

I'm part way through implementing password_auth
I've switched password to be a string rather than [u8] as that seems to be simpler with verify_password.

Now I think I need to generate_hash for the the creds_password (what comes from the login form).

So the generate_hash of the login password should match the stored hash of the password in the user.

BUT generate_hash seems to be giving a different value each time it is called eg

```
            let p = generate_hash(&creds.password);
            let p2 = generate_hash(&creds.password);
            let p3 = generate_hash(&creds.password);
            let p4 = generate_hash(&creds.password);
            println!("authenticate verify_password p     :{}", p);
            println!("authenticate verify_password p2    :{}", p2);
            println!("authenticate verify_password p3    :{}", p3);
            println!("authenticate verify_password p4    :{}", p4);
```

You get 4 different passwords. So how would this ever match what we have stored earlier in the User?

**Comment by maxcountryman** [comment]
I want to make sure we're tracking toward a direction that meets the general use case of this crate: currently we have no opinion about what kind of credentials an application uses and in fact support flows without passwords (see [oauth](https://github.com/maxcountryman/axum-login/tree/main/examples/oauth2)).

@czocher does it make sense to include such an implementation when some applications won't use it? My inclination is that it does not and instead we should provide this as an example.

**Reply by dave42w** [reply]
I agree that we shouldn't impose and that would be my end goal.

**Reply by czocher** [reply]
I agree @maxcountryman, we should rather provide examples than implement everything as part of the library.

**Comment by czocher** [comment]
@dave42w You shouldn't compare two hashed passwords, instead you just need the hashed version from the database and the plaintext version (from the user login form) and use: https://docs.rs/password-auth/latest/password_auth/fn.verify_password.html

This function takes the hash + the plaintext and verifies if they're both equal.

Also FYI, a password hashing function usually produces a string with a couple of fields in PHC standardized form: https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md
You're getting different results from each hash generation because the salt is always randomly generated and different for each attempt.

What a password verification function does is take the salt from the hashed password saved in the database, rehash the provided plaintext password using that salt and then compare them.

As a thought experiment we can imagine a password hashing algorithm that always produces the same result (so either the same salt or no salt at all). If that algorithm was used on a big enough database, you could easily check which users have the same password by comparing the hashes together. The salt makes it a lot harder.

**Reply by dave42w** [reply]
Thanks, very helpful.
I'll restart my fork to make sure it is cleanly separated into the example (apart from minimal replace `ring` with minimum needed from `password-auth`).
Dave
